api/client.go

```diff
@@ -1145,6 +1151,103 @@ func RequireState(states ...string) RequestCallback {
 	}
 }
 
+// CompareReplicationStates returns 1 if s1 is newer or identical, -1 if s1 is older, and 0
+// if neither s1 or s2 is strictly greater. An error is returned if s1 or s2
+// are invalid or from different clusters.
+func CompareReplicationStates(s1, s2 string) (int, error) {
+	w1, err := ParseRequiredState(s1, nil)
+	if err != nil {
+		return 0, err
+	}
+	w2, err := ParseRequiredState(s2, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	if w1.ClusterID != w2.ClusterID {
+		return 0, fmt.Errorf("can't compare replication states with different ClusterIDs")
+	}
+
+	switch {
+	case w1.LocalIndex >= w2.LocalIndex && w1.ReplicatedIndex >= w2.ReplicatedIndex:
+		return 1, nil
+	// We've already handled the case where both are equal above, so really we're
+	// asking here if one or both are lesser.
+	case w1.LocalIndex <= w2.LocalIndex && w1.ReplicatedIndex <= w2.ReplicatedIndex:
+		return -1, nil
+	}
+
+	return 0, nil
+}
+
+func MergeStates(old []string, new string) []string {
+	if len(old) == 0 || len(old) > 2 {
+		return []string{new}
+	}
+
+	var ret []string
+	for _, o := range old {
+		c, err := CompareReplicationStates(o, new)
+		if err != nil {
+			return []string{new}
+		}
+		switch c {
+		case 1:
+			ret = append(ret, o)
+		case -1:
+			ret = append(ret, new)
+		case 0:
+			ret = append(ret, o, new)
+		}
+	}
+	return strutil.RemoveDuplicates(ret, false)
+}
+
+func ParseRequiredState(raw string, hmacKey []byte) (*logical.WALState, error) {
```

Comment:
```suggestion
func ParseReplicationState(raw string, hmacKey []byte) (*logical.WALState, error) {
```

Also, does this one need to be public? I don't see it or `MergeStates` used in your TF provider PR.

Reply:
Ah this was public when it was in Core and I didn't give a second deeper thought. We don't need `ParseReplicationState` to be public, but `MergeStates` does need to be since it is being used here https://github.com/hashicorp/terraform-provider-vault/pull/1170/files#diff-7e69de666c9006fa0a0fb8fe4e4f8c418acce7c555f69f8087d88d16b434f2c1R714

Reply:
Oh good point, I should have said `CompareReplicationStates` instead of `MergeStates`

Reply:
Updated and made necessary methods private